notify/src/address/tracker.rs

```diff
@@ -0,0 +1,649 @@
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.0.capacity()
+    }
+}
+
+impl Indexer for CounterMap {
+    fn contains(&self, index: Index) -> bool {
+        self.0.contains_key(&index)
+    }
+
+    fn insert(&mut self, index: Index) -> bool {
+        let mut result = true;
+        self.0
+            .entry(index)
+            .and_modify(|x| {
+                *x += 1;
+                result = *x == 1;
+            })
+            .or_insert(1);
+        result
+    }
+
+    fn remove(&mut self, index: Index) -> bool {
+        let mut result = false;
+        self.0.entry(index).and_modify(|x| {
+            if *x > 0 {
+                *x -= 1;
+                result = *x == 0
+            }
+        });
+        result
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+#[cfg(test)]
+#[derive(Debug, Clone)]
+#[cfg(test)]
+pub struct Counter {
+    pub index: Index,
+    pub count: RefCount,
+    pub locked: bool,
+}
+
+#[cfg(test)]
+impl Counter {
+    pub fn new(index: Index, count: RefCount) -> Self {
+        Self { index, count, locked: false }
+    }
+
+    pub fn active(&self) -> bool {
+        self.count > 0
+    }
+}
+
+#[cfg(test)]
+impl PartialEq for Counter {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index
+    }
+}
+#[cfg(test)]
+impl Eq for Counter {}
+
+#[cfg(test)]
+impl PartialOrd for Counter {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+#[cfg(test)]
+impl Ord for Counter {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.index.cmp(&other.index)
+    }
+}
+
+/// Set of `Index`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IndexSet(HashSet<Index>);
+
+impl IndexSet {
+    pub fn new(indexes: Vec<Index>) -> Self {
+        Self(indexes.into_iter().collect())
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(HashSet::with_capacity(capacity))
+    }
+
+    pub fn iter(&self) -> hash_set::Iter<'_, Index> {
+        self.0.iter()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.0.capacity()
+    }
+
+    pub fn drain(&mut self) -> hash_set::Drain<'_, Index> {
+        self.0.drain()
+    }
+}
+
+impl Indexer for IndexSet {
+    fn contains(&self, index: Index) -> bool {
+        self.0.contains(&index)
+    }
+
+    fn insert(&mut self, index: Index) -> bool {
+        self.0.insert(index)
+    }
+
+    fn remove(&mut self, index: Index) -> bool {
+        self.0.remove(&index)
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+#[derive(Debug)]
+struct Inner {
+    script_pub_keys: IndexMap<ScriptPublicKey, RefCount>,
```

Comment:
I would add a noticeable comment/warning here emphasizing that every mutable access to `script_pub_keys` must be careful to not use `IndexMap` APIs which alter the index.